src/v/cluster/tx_gateway.cc

```diff
@@ -56,24 +55,35 @@ tx_gateway::abort_tx(abort_tx_request&&, rpc::streaming_context&) {
     return ss::make_ready_future<abort_tx_reply>(abort_tx_reply());
 }
 
-ss::future<begin_group_tx_reply>
-tx_gateway::begin_group_tx(begin_group_tx_request&&, rpc::streaming_context&) {
-    co_return begin_group_tx_reply();
+ss::future<begin_group_tx_reply> tx_gateway::begin_group_tx(
+  begin_group_tx_request&& request, rpc::streaming_context&) {
+    return _rm_group_proxy->begin_group_tx_locally(
+      std::move(request.group_id), request.pid, request.timeout);
 };
 
 ss::future<prepare_group_tx_reply> tx_gateway::prepare_group_tx(
-  prepare_group_tx_request&&, rpc::streaming_context&) {
-    co_return prepare_group_tx_reply();
+  prepare_group_tx_request&& request, rpc::streaming_context&) {
+    return _rm_group_proxy->prepare_group_tx_locally(
+      std::move(request.group_id),
+      request.etag,
+      request.pid,
+      request.tx_seq,
+      request.timeout);
 };
 
 ss::future<commit_group_tx_reply> tx_gateway::commit_group_tx(
-  commit_group_tx_request&&, rpc::streaming_context&) {
-    co_return commit_group_tx_reply();
+  commit_group_tx_request&& request, rpc::streaming_context&) {
+    return _rm_group_proxy->commit_group_tx_locally(
```

Comment:
are there safety mechanisms in place in the `*_locally(...)` variants that deal with race conditions where an operation is dispatched over the network, but the state of the system has changed in the mean time such that the receiving node is no longer the correct node to handle the operation?

Reply:
this concern is always valid because a system may change leadership after the check has passed. here in the transactional methods of group.cc we control it by avoiding mutating important state before replication and when we do conditional replication to avoid ABA problem